File: KT17/WithinMP.py
import numpy as np
from ._CFunctions import _CWithinMP

def WithinMP(x, y, z, Rsm):
	'''
	Tests whether a position in MSM coordinates is within the 
	magnetopause.
	
	Inputs
	======
	x : float
		Scalar or array of x MSM coordinate(s) (Rm)
	y : float
		Scalar or array of y MSM coordinate(s) (Rm)
	z : float
		Scalar or array of z MSM coordinate(s) (Rm)
	Rsm : float
		Subsolar magnetopause standoff distance (Rm).
			
	Returns
	=======
	res : bool
		True if within the magnetopause.

	'''
	#Convert input variables to appropriate numpy dtype:
	if np.size(x) == 1:
		_x = np.float64(x)
		_y = np.float64(y)
		_z = np.float64(z)
		_Rsm = np.float64(Rsm)
		res = _CWithinMP(_x, _y, _z, _Rsm)
	else:
		n = np.size(x)
		res = np.zeros(n,dtype='bool')
		rsm = np.zeros(n,dtype='float64') + Rsm
		for i in range(0,n):
			_x = np.float64(x[i])
			_y = np.float64(y[i])
			_z = np.float64(z[i])
			_Rsm = np.float64(rsm[i])
			res[i] = _CWithinMP(_x, _y, _z, _Rsm)			

	return res

File: KT17/__data/libkt/magnetopause/withinmp.cc
#include "withinmp.h"

/***********************************************************************
 * NAME : bool WithinMP(r,theta,Rsm)
 * 
 * DESCRIPTION : Checks whether a position is within the magnetopause.
 * 
 * INPUTS : 
 * 		double r		Radial coordinate.
 * 		double theta	Angle between y/z and x coordinates (rad).
 * 		double Rsm		Subsolar magnetopause distance (Rm).
 *
 * RETURNS :
 *		bool wmp		true if within magnetopause.
 * 
 * ********************************************************************/
bool WithinMPRT(double r, double theta, double Rsm) {
	
	double rmp;
	
	rmp = MagnetopauseR(theta,Rsm);
	
	return (r <= rmp);
	
}

/***********************************************************************
 * NAME : bool WithinMP(x,y,z,Rsm)
 * 
 * DESCRIPTION : Checks whether a position is within the magnetopause.
 * 
 * INPUTS : 
 * 		double x		x-coordinate (Rm).
 * 		double y		y-coordinate (Rm).
 * 		double z		z-coordinate (Rm).
 * 		double Rsm		Subsolar magnetopause distance (Rm).
 *
 * RETURNS :
 *		bool wmp		true if within magnetopause.
 * 
 * ********************************************************************/
bool WithinMP(double x, double y, double z, double Rsm) {
	
	double theta, r, rho;
	rho = y*y + z*z;
	r = x*x + rho;
	rho = sqrt(rho);
	r = sqrt(r);
	theta = atan2(rho,x);
	
	return WithinMPRT(r,theta,Rsm);

}
